### PR TITLE
feat: Flagd web provider

### PR DESCRIPTION
**Pull request by james-milligan**

### CODE REVIEW

File: .release-please-manifest.json [hunk 1]
```diff
@@ -1,5 +1,7 @@
 {
   "libs/hooks/open-telemetry": "3.0.0",
   "libs/providers/go-feature-flag": "0.3.0",
-  "libs/providers/flagd": "0.5.1"
+  "libs/providers/flagd": "0.5.1",
+  "libs/providers/flagd-web": "0.1.0",
+  "libs/providers/web-flagd": "0.1.0"
```
Comment (beeme1mr) [inline .release-please-manifest.json:6]: Is this valid?

File: libs/providers/flagd-web/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# NodeJS flagd-web Provider for OpenFeature
```
Comment (beeme1mr) [inline libs/providers/flagd-web/README.md:1]:
```suggestion
# Client-side flagD Provider for OpenFeature
```

File: libs/providers/flagd-web/README.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+![Experimental](https://img.shields.io/badge/experimental-breaking%20changes%20allowed-yellow)
+
+## Installation
+
+```
+$ npm install @openfeature/flagd-web-provider
+```
```
Comment (beeme1mr) [inline libs/providers/flagd-web/README.md:9]: Are any peer dependencies required?

File: package.json [hunk 4]
```diff
@@ -11,10 +11,13 @@
   },
   "private": true,
   "dependencies": {
+    "@bufbuild/connect-web": "^0.2.1",
     "@grpc/grpc-js": "^1.6.7",
+    "@openfeature/nodejs-sdk": "^0.3.2",
```
Comment (beeme1mr) [inline package.json:16]: Please rebase and use `@openfeature/js-sdk` which should be available as a dev dependency

File: package.json [hunk 5]
```diff
@@ -11,10 +11,13 @@
...
     "@opentelemetry/api": "^1.1.0",
     "@protobuf-ts/grpc-transport": "^2.7.0",
     "axios": "^0.27.2",
+    "jest-fetch-mock": "^3.0.3",
```
Comment (beeme1mr) [inline package.json:20]: Move the dev dependencies

File: release-please-config.json [hunk 6]
```diff
@@ -24,6 +22,20 @@
...
       "versioning": "default"
+    },
+    "libs/providers/flagd-web": {
+      "release-type": "node",
+      "prerelease": true,
+      "bump-minor-pre-major": true,
+      "bump-patch-for-minor-pre-major": true,
+      "versioning": "default"
+    },
+    "libs/providers/web-flagd": {
```
Comment (beeme1mr) [inline release-please-config.json:33]: I think you can remove this.

File: tsconfig.base.json [hunk 7]
```diff
@@ -10,27 +10,25 @@
...
+      "@openfeature/flagd-web-provider": [
+        "libs/providers/flagd-web/src/index.ts"
       ],
       "@openfeature/go-feature-flag-provider": [
         "libs/providers/go-feature-flag/src/index.ts"
       ],
       "@openfeature/hooks-open-telemetry": [
         "libs/hooks/open-telemetry/src/index.ts"
+      ],
+      "@openfeature/web-flagd-provider": [
```
Comment (beeme1mr) [inline tsconfig.base.json:28]: I think you can remove this.

**Comment by agardnerIT** [comment]
How would I use this (I'm not good with Javascript but want to test this out) with Javascript?

I have flagd running on `8013` and can:

```
curl -X POST http://example.com:8013/schema.v1.Service/ResolveString -H "Content-Type: application/json" -d '{"flagKey": "foo", "context": {}}'
```
Which returns:
```
{"value":"bar","reason":"DEFAULT","variant":"key"}
```

Then I have: `npm install @openfeature/flagd-provider`

Now I imagine I need to do something like this:

```
<!DOCTYPE html>
<html>
<head>
  <title>Dummy Page</title>
  <script type="module">
    import OpenFeature from 'openfeature/js-sdk';
    OpenFeature.setProvider(new FlagdProvider({ host: 'example.com', port: '8013' }));

    const client = OpenFeature.getClient('my-client');
    const value = client.getStringValue('foo', false);

    // "Value is: bar"
    console.log("Value is: ", value); 
  </script>
</head>
<body>
  <h1>test page</h1>
  <p>This is the test page...To be implemented.</p>
</body>
</html>
```

I'm new to OpenFeature so perhaps I'm way off here!

File: .release-please-manifest.json (hunk 1 above)
Comment (james-milligan) [reply .release-please-manifest.json]: removed

File: package.json (hunk 4 above)
Comment (james-milligan) [reply package.json]: updated

File: package.json (hunk 5 above)
Comment (james-milligan) [reply package.json]: moved

File: release-please-config.json (hunk 6 above)
Comment (james-milligan) [reply release-please-config.json]: removed

File: tsconfig.base.json (hunk 7 above)
Comment (james-milligan) [reply tsconfig.base.json]: removed

**Comment by james-milligan** [comment]
requires https://github.com/open-feature/js-sdk/pull/200

to test use the current experiemental sdk version: 

`@openfeature/js-sdk@0.4.0-experimental-e9ba70956ed77c7b41032b149a8f6bf915ae50e4`

**Comment by james-milligan** [comment]
> How would I use this (I'm not good with Javascript but want to test this out) with Javascript?

For now as the experimental version of the sdk is required the easiest way to test this is to clone the repo, update the sdk version to match the above comment, then install the package via filepath. 

1. Clone repo
2. Update the js-sdk version in `package.json` to `0.4.0-experimental-e9ba70956ed77c7b41032b149a8f6bf915ae50e4`
3. Run the package command from `libs/providers/flagd-web/project.json`

now you can run `npm install PATH/TO/CLONED/REPO/dist/libs/providers/flagd-web` to import the package as `@openfeature/flagd-web-provider` in your above example, replacing `new FlagdProvider` with `new FlagdWebProvider`

let me know if you have any issues (i will update this comment once the linked pr is merged and the sdk is updated for this PR)

**Comment by agardnerIT** [comment]
Thanks James, I feel like I'm closer. Here's an entire log of what I did.

```
cd c:/path/to/folder/flagd-tests
git clone https://github.com/open-feature/js-sdk-contrib
cd js-sdk-contrib
git pull origin pull/112/head
npm install

1 critical package found

npm audit fix
npm run package
> js-sdk-contrib@0.0.0 package
> npx nx run-many --all --target=package
    √  nx run hooks-open-telemetry:package (18s)
    √  nx run providers-go-feature-flag:package (28s)
    ×  nx run providers-flagd:generate
       Submodule 'libs/providers/flagd-web/schemas' (https://github.com/open-feature/schemas.git) registered for path '../flagd-web/schemas'
Submodule 'libs/providers/flagd/schemas' (https://github.com/open-feature/schemas) registered for path 'schemas'
Cloning into 'C:/path/to/folder/flagd-tests/js-sdk-contrib/libs/providers/flagd-web/schemas'...
Cloning into 'C:/path/to/folder/flagd-tests/js-sdk-contrib/libs/providers/flagd/schemas'...
From https://github.com/open-feature/schemas
 * branch            02945d277d1caa361ff9babe20653f98860f3778 -> FETCH_HEAD
Submodule path '../flagd-web/schemas': checked out '02945d277d1caa361ff9babe20653f98860f3778'
Submodule path 'schemas': checked out '185e3bd777755efcca3cb5d7471c8fe27154e496'
'buf' is not recognized as an internal or external command,
       operable program or batch file.
       ERROR: Something went wrong in run-commands - Command failed: cd schemas && buf generate buf.build/open-feature/flagd --template protobuf/buf.gen.ts.yaml
    ×  nx run providers-flagd-web:generate
       Submodule path '../flagd/schemas': checked out '185e3bd777755efcca3cb5d7471c8fe27154e496'
'buf' is not recognized as an internal or external command,
       operable program or batch file.
       ERROR: Something went wrong in run-commands - Command failed: cd schemas && buf generate buf.build/open-feature/flagd --template protobuf/buf.gen.ts-connect.yaml                                                                                                                                                                                                            
 ————————————————————————————————————————————————————————
                                                                                                                                                                                                                         
 >  NX   Ran target package for 4 projects and 2 task(s) they depend on (42s)                                                                                                                                            
                                                                                                                                                                                                                         
    √    2/4 succeeded [0 read from cache]                                                                                                                                                                               
                                                                                                                                                                                                                         
    ×    2/4 targets failed, including the following:                                                                                                                                                                    
         - nx run providers-flagd:generate                                                                                                                                                                               
         - nx run providers-flagd-web:generate

// Get buf CLI from https://github.com/bufbuild/buf/releases/tag/v1.8.0 - add to PATH
Close VisualStudio code (buf isn't recognised even on new terminal windows for some reason) then:
buf (to ensure it works)
cd c:\path\to\folder\flagd-tests\js-sdk-contrib
npm run package

npm install c:\path\to\folder\flagd-tests\js-sdk-contrib\dist\libs\providers\flagd-web

So far so good (no more errors)...

CORS blocked shown in chrome console because we can't use file:\\ so install "Live Server" from VS Code extensions
# See also: https://stackoverflow.com/questions/52919331/access-to-script-at-from-origin-null-has-been-blocked-by-cors-policy

Change index.html include to:

```
import OpenFeature from './js-sdk-contrib/dist/libs/providers/flagd-web/index.js'
```

In chrome Console:

```
Uncaught TypeError: Failed to resolve module specifier "@openfeature/js-sdk". Relative references must start with either "/", "./", or "../".
```
```

Remember I'm using pure JavaScript, not NodeJS.

File: libs/providers/flagd-web/README.md [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+## Building
```
Comment (beeme1mr) [inline libs/providers/flagd-web/README.md:11]: Add section on the git submodule and Buf requirements. I believe you already have something similar in the server-side provider.

**Comment by beeme1mr** [comment]
@agardnerIT Buf is required to generate the code. This will happen automatically within the GitHub Action once the package is officially released.

**Comment by james-milligan** [comment]
> Thanks James, I feel like I'm closer. Here's an entire log of what I did.

Once the buf issue is resolved following the comment from @beeme1mr, you should be able to get the test working by updating the code as seen below:
```
import OpenFeature from 'openfeature/js-sdk';
import FlagdWebProvider from '@openfeature/flagd-web-provider'
OpenFeature.setProvider(new FlagdWebProvider({ host: 'example.com', port: '8013' }));

const client = OpenFeature.getClient('my-client');
const value = client.getStringValue('foo', false);
```

Let me know if this helps, it may be worthwhile pushing your code to a repo so i can take a look if there are any other issues!

File: libs/providers/flagd-web/src/lib/flagd-web-provider.ts [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+function ErrorResponse(err: unknown): string {
+  err as Partial<ConnectError>
+  switch ((err as Partial<ConnectError>).code) {
+    case Code.NotFound:
+      return ErrorCode.FLAG_NOT_FOUND
+    case Code.InvalidArgument:
+      return ErrorCode.TYPE_MISMATCH
+    case Code.Unavailable:
+        return ERROR_DISABLED
```
Comment (toddbaert) [inline libs/providers/flagd-web/src/lib/flagd-web-provider.ts:156]: I think we may be mixing reasons/error semantics here.

I don't think disabled flag actually constitutes an error. This is evidenced by the fact it's not an enumerate error code, and it's really not an unexpected condition. A flag being disabled is something somebody chose to do, I don't think that logically it's an error condition. It's a `reason`.

cc @beeme1mr

File: libs/providers/flagd-web/src/lib/flagd-web-provider.ts [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    case Code.DataLoss:
+        return ERROR_PARSE_ERROR
```
Comment (toddbaert) [inline libs/providers/flagd-web/src/lib/flagd-web-provider.ts:158]: We have a PARSE_ERROR error code.

**Comment by toddbaert** [comment]
I'll bring this one home as we discussed @james-milligan . Thanks for this, it's great!

**Comment by toddbaert** [comment]
Closing in favor of: https://github.com/open-feature/js-sdk-contrib/pull/142

**Pull request closed without merging**